File: src/w3hw_ros2py/w3hw_ros2py/subscriber_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from w3hw_ros2py_interfaces.srv import AddTwoInts

class HelloWorldSubscriber(Node):
    def __init__(self):
        super().__init__('hello_world_subscriber')

        # 创建订阅者，订阅 "hello_world" 话题
        self.subscription = self.create_subscription(
            String,
            'hello_world',
            self.listener_callback,
            10)

        self.get_logger().info('HelloWorld订阅者已启动，等待接收消息...')

    def listener_callback(self, msg):
        # 当收到消息时，输出接收到的词条
        self.get_logger().info(f'收到消息: "{msg.data}"')

def main(args=None):
    rclpy.init(args=args)
    node = HelloWorldSubscriber()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/w3hw_ros2cpp/src/subscriber_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class HelloWorldSubscriber : public rclcpp::Node
{
public:
    HelloWorldSubscriber() : Node("hello_world_subscriber")
    {   
        //创建订阅者，订阅 "hello_world" 话题
        subscription_ = this->create_subscription<std_msgs::msg::String>(
            "hello_world", 10,
            std::bind(&HelloWorldSubscriber::topic_callback, this, std::placeholders::_1));
        RCLCPP_INFO(this->get_logger(), "HelloWorld订阅者已启动，等待消息...");
    }

private:
    void topic_callback(const std_msgs::msg::String::SharedPtr msg) const
    {   
        //当收到消息时，输出接收到的词条
        RCLCPP_INFO(this->get_logger(), "收到消息: \"%s\"", msg->data.c_str());
    }

    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<HelloWorldSubscriber>());
    rclcpp::shutdown();
    return 0;
}
